draft-ietf-quic-transport.md

```diff
@@ -1741,6 +1741,13 @@ parameters as a connection error of type TRANSPORT_PARAMETER_ERROR.
 Endpoints use transport parameters to authenticate the negotiation of
 connection IDs during the handshake; see {{cid-auth}}.
 
+When sent by the client, transport parameters are not specific to a particular
+ALPN being negotiated.  Therefore, a client cannot have application protocol
+specific values in the transport parameters.  Application protocols should take
+this into account when designing versioning schemes, and not place restrictions
+on the values in the transport parameters that might prevent protocol changes
+in future version of the application protocol.
```

Comment:
Regarding the last sentence, how about saying something like: _Application protocols SHOULD NOT introduce prohibitive requirements that prevent other application protocols from working._

I am not sure if this problem can be scoped to different versions of one protocol. While that could be the case for HTTP, there could be cases where a client might ask for using different protocols that provide the same service. To give an example, a client might send send an ALPN of ("h3" (for doing DNS over HTTP), "dns" (for doing DNS over QUIC)) to a DNS server.

Reply:
@kazuho I'm not sure how you'd go about not making _prohibitive requirements_ for other protocols without understanding the set of all other possible protocols **or** by setting no requirements at all.

Reply:
@nibanks If that's unclear, saying something like the following might be better: "should not introduce restrictions to the transport parameters, as doing so might prevent clients from offering multiple application protocols."

I think that addresses the two examples that you laid out in the original comment of #3965.

Reply:
There is a difference between requirements and recommendations.  We do recommend in HTTP/3 that a certain number of unidirectional stream be available, but it is [just a recommendation](https://quicwg.org/base-drafts/draft-ietf-quic-http.html#section-6.2-4).

I think that this could be clearer then:

> Protocol negotiation with Application Layer Protocol Negotiation (ALPN; see {{?ALPN=RFC7301}}) allows clients to offer multiple application protocols during connection establishing. The single set of transport parameters offered by a client apply to all application protocols that the client offers.  Application protocols can make recommendations about the values for transport parameters, such as suggesting minimum values for stream or flow control limits.  Application protocols that place hard constraints on transport parameters could make it impossible for clients to attempt a connection that includes other application protocols with different constraints.

Reply:
I'd like to make some editorial suggestions to martin's proposal, but I like that text.